src/goto-programs/abstract-interpretation/gcse.h

```diff
@@ -0,0 +1,167 @@
+#pragma once
+
+// GCSE - Global Common Subexpression Elimination
+
+#include <util/message.h>
+#include <goto-programs/abstract-interpretation/ai.h>
+#include <pointer-analysis/value_set_analysis.h>
+/**
+ * @brief Abstract domain to obtain all available expressions (AE)
+ *
+ * The domain is the set of expressions that were already computed up to this point.
+ *
+ * ... // AE: []
+ * int x = a + b + c // AE: [a + b, a + b + c]
+ * int y = a + b + d // AE: [a + b, a + b + c, a + b + d]
+ * c = 42 // AE: [a + b, a + b + d]
```

Comment:
Is `a + b + c` missing for this last step?

Reply:
It is not, `a+b+c` is not available anymore because `c` changed. So the expression would need to be recomputed again

Reply:
Right, that makes sense. Thanks.